### PR TITLE
<calcite-tree> component

### PR DESCRIPTION
**Pull request by patrickarlt**
This is an initial take on `<calcite-tree>` component for https://github.com/Esri/calcite-components/issues/11. Right now everything is basically setup and working. I might rework a lot of this to rely less on DOM querying and do everything with events and a registry like I did for tabs but I wanted to get feedback on styling and use cases before I dive into right-to-left and accessibility.

Right now this:

* Supports a light and dark theme
* Infinite nesting
* Optional lines
* Optional "selection" of items
   * Single item selected
   * Multiple items selected
   * Single item or sub-tree selected (VTSE style)
   * Multiple items or sub-trees selected (VTSE style)

I still need to add an event for when the selection changes but this should allow this to be used in both VTSE and the developers site for tree navigation.

@macandcheese this should work for the ArcGIS Online filter trees you posted in https://github.com/Esri/calcite-components/issues/11#issuecomment-520521931 but I'm wondering if that would be better as a separate component. Right now I think you would have to create special styles for radio buttons for checkboxes to allow the radio button or checkbox to control the active state as opposed to the `<calcite-tree-item>`. Let me know what you think.

Design wise I'm a little worried about the horizontal width in VTSE. I think the dimensions might be pretty good otherwise. I'm wondering if in "compact" mode we really need the circle indicators? If compact mode could just drop the circle indicators I think we would be good but does that change the overall feel of the component?

### CODE REVIEW

**Comment by macandcheese** [comment]
Awesome! This is looking good - I think all the functionality is there. 

Some nitpicky style things but overall this is a great start. Re: density - I think "compact" prop can reduce the x-padding as well as y-padding by .25rem or so for each "level" that could alleviate that, and perhaps we can play around with removing the circles or at least the x-padding there as well if that's not enough. I can take a stab at some of the styling stuff this week if you are tied up. 

Good stuff!

**Comment by macandcheese** [comment]
Did a little playing around, I think we can make the "scale / density" prop work with the active dot. Just did this quick so it'd be a little cleaner bu, something like... 

Artboard Copy

**Comment by bstifle** [comment]
dig it!

**Comment by patrickarlt** [comment]
@macandcheese I added a compact mode which I worked on yesterday. I think it looks pretty close to your designs and I agree that we should keep the indicator. I've also added proper ARIA tags.

I think the only things that I would like to do before merging this would be LTR and key board navigation and I can follow up with tests.

**Comment by patrickarlt** [comment]
Added a few more things:

* RTL support
* better hover states
* Differentiated hover/focus vs selected states (used `h-bb-40` instead of `ui-blue-pressed`)
* Added a `calciteTreeSelect` event which emits the selected nodes.

Going to try to wrap up keyboard navigation today.

**Comment by patrickarlt** [comment]
@macandcheese @paulcpederson I think this is ready for a review. I would like to add some unit tests but I would also like to add unit tests but I would also like to cut a release and start using this in some projects ASAP.

File: src/components/calcite-tree/calcite-tree.tsx [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  //--------------------------------------------------------------------------
+  //
+  //  Event Listeners
+  //
+  //--------------------------------------------------------------------------
+
+  @Listen("focus") onFocus(e) {
+    console.log("focus from", e.target, this.root);
```
Comment (macandcheese) [inline src/components/calcite-tree/calcite-tree.tsx]: Stray console

**Comment by patrickarlt** [comment]
@macandcheese  I added the nested lines. They don't show as selected (not really sure how to accomplish that) and this now resolves the tab issues in https://github.com/Esri/calcite-components/issues/100.

**Comment by patrickarlt** [comment]
Also keyboard nav is done. @macandcheese I'm not sure we can/want to resolve the double focus thing. Technically there is an `<a>` inside the `<calcite-tree-item>` By default if you "activate" the tree item with <kbd>Enter</kbd> (the recommended keyboard interaction for activating a tree node) then it will try to find a link and click it.

The default of using <kbd>Tab</kbd> is to just select the "next focus-able thing" so I think this behavior is correct. This also allows uses to switch between keyboard nav for the tree and keyboard nav on the individual links. I'm also pretty sure that `tabindex="-1"` on the link will disable this but the default action on the tree item will still work.

**Comment by macandcheese** [comment]
Ok, I think that makes sense. Sounds good 🤙

**Comment by paulcpederson** [comment]
looks dope, let's get 'er in the ol' master line

**Pull request merged**